### PR TITLE
ENH: Add support for median averaging and no averaging to Welch PSD

### LINKED ISSUE #3336: mne.time_frequency.psd_welch add median option
I want to robustly estimate a power spectrum from raw data. 
I am concerned that segments of my raw data contain artifacts. 
mne.time_frequency.psd_welch divides the raw data into N (possibly overlapping) windows, computes N psds, and returns the average psd. 
Could we add the option to return the median psd? 
By this I mean for each frequency bin we have N estimates of the power (from the N psds), and we return the median of these N values. 
I believe the R function welchPSD does this.

### PR DESCRIPTION
**Pull request by hoechenberger**
#### Reference issue
Closes #3336.

#### What does this implement/fix?

- Adds an `average` kwarg to `psd_welch()` and `psd_array_welch()`:
  - `average='mean'`: returns PSD based on the arithmetic mean of all segments, calculated via `scipy.signal.welch()`
  - `average='median'`: returns PSD based on the median of all segments, corrected for bias relative to the mean, calculated via `scipy.signal.welch()`
  - `average=None`: returns all segments unaggregated, calculated via `scipy.signal.spectrogram()`

- <strike>There is currently quite some code duplication in the tests (one for `Raw`, one for `Epochs`). I didn’t want to throw these tests together into a single test function, though, to keep complexity low.</strike>

- <strike>I’ve renamed the auxiliary function `_psd_func()` to `_spect_func()` (calculates spectrograms), and added yet another function `_welch_func()` (calculates Welch PSD). The sole reason these functions exist is to help with parallelization via `parallel_func()`. I’m not sure if they couldn’t be squeezed into an “inline-lambda” too, as re-usability honestly seems quite limited anyway…</strike>

-  Parallelization (`n_jobs > 1`) is currently broken! I’ve tried a while to get to the bottom of this, but this seems to be broken in `master` already and I don’t really know how to debug it: my debugger would only step into the first thread (or is it a process?), which completes successfully; the second, however, would raise an exception. Would probably need to set up a remote debugger.

- <strike>I’ve updated the docstrings, but no other parts of the documentation so far.</strike>

- <strike>One CI job fails because the installed version of SciPy is too old.</strike>

- <strike>Changelog entry still missing.</strike>

MWE:
```python
import numpy as np
import os
import mne


base_dir = '/Users/hoechenberger/Development/mne-python/mne/io/tests/data'
raw_fname = os.path.join(base_dir, 'test_raw.fif')

rng = np.random.RandomState(40)
raw = mne.io.read_raw_fif(raw_fname)
data = 0.1 * rng.randn(len(raw.ch_names), raw.n_times)
raw = mne.io.RawArray(data, raw.info)

fmin, fmax = 20, 100
n_fft = 256
n_per_seg = 128

kws = dict(fmin=fmin,
           fmax=fmax,
           n_fft=n_fft,
           n_per_seg=n_per_seg)

psds_mean, freqs = mne.time_frequency.psd_welch(raw, average='mean', **kws)
psds_median, freqs = mne.time_frequency.psd_welch(raw, average='median', **kws)


import matplotlib.pyplot as plt
plt.plot(freqs, psds_mean[0], label='mean')
plt.plot(freqs, psds_median[0], label='median')
plt.legend()
plt.title('PSD')
plt.xlabel('Freq (Hz)')
plt.ylabel('Power (V^2 / Hz)')
plt.show()
```
PSD

cc @agramfort @larsoner @mmagnuski

### CODE REVIEW

**Comment by larsoner** [comment]
> There is currently quite some code duplication in the tests (one for Raw, one for Epochs). I didn’t want to throw these tests together into a single test function, though, to keep complexity low.

You can use a `pytest` decorator like:
```
@pytest.mark.parametrize('kind', ('raw', 'epochs'))
def test_whatever(kind):
    if kind == 'raw':
        raw = ...
    else:
        assert kind == 'epochs'
        epochs = ...
    ...
```
It cuts down on code dup.

> I’m not sure if they couldn’t be squeezed into an “inline-lambda” too, as re-usability honestly seems quite limited anyway…

`lambda` sometimes makes things cleaner, but this is probably not one of those times.

> I’ve updated the docstrings, but no other parts of the documentation so far.

You could push to some `examples` or `tutorials` file to increase exposure.

> One CI job fails because the installed version of SciPy is too old.

See how we use `requires_version` decorator

**Comment by larsoner** [comment]
(I will wait to review until you think it's ready for merge)

File: mne/time_frequency/psd.py [hunk 1]
```diff
@@ -160,6 +194,11 @@ def psd_welch(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None, n_fft=256,
     n_per_seg : int | None
         Length of each Welch segment (windowed with a Hamming window). Defaults
         to None, which sets n_per_seg equal to n_fft.
+    average : str | None
+        How to average  to segments. If ``mean`` (default), calculate the
```
Comment (agramfort) [inline mne/time_frequency/psd.py]:
```suggestion
        How to average to segments. If ``mean`` (default), calculate the
```

File: mne/time_frequency/psd.py [hunk 2]
```diff
@@ -160,6 +194,11 @@ def psd_welch(inst, fmin=0, fmax=np.inf, tmin=None, tmax=None, n_fft=256,
...
+        arithmetic mean. If ``median``, calculate the median, corrected for
+        its bias relative to thee mean. If ``None``, returns the unaggregated
```
Comment (agramfort) [inline mne/time_frequency/psd.py]:
```suggestion
        its bias relative to the mean. If ``None``, returns the unaggregated
```

File: mne/time_frequency/tests/test_psd.py [hunk 3]
```diff
@@ -160,6 +160,116 @@ def test_psd():
...
+    n_overlap = 0
+
+    kws = dict(fmin=fmin,
+               fmax=fmax,
+               tmin=tmin,
+               tmax=tmax,
+               n_fft=n_fft,
+               n_per_seg=n_per_seg,
+               n_overlap=n_overlap,
+               picks=picks_psd)
```
Comment (agramfort) [inline mne/time_frequency/tests/test_psd.py]: put all this on 2 lines.

File: mne/time_frequency/tests/test_psd.py [hunk 4]
```diff
@@ -160,6 +160,116 @@ def test_psd():
...
+    # Populate raw with sinusoids
+    rng = np.random.RandomState(40)
+    data = 0.1 * rng.randn(len(raw.ch_names), raw.n_times)
+    freqs_sig = [8., 50.]
+    for ix, freq in zip(picks_psd, freqs_sig):
+        data[ix, :] += 2 * np.sin(np.pi * 2. * freq * raw.times)
+
+    raw = RawArray(data, raw.info)
+
+    tmin, tmax = 0, 20  # use a few seconds of data
```
Comment (agramfort) [inline mne/time_frequency/tests/test_psd.py]: how long does it take to test? tests should be as fast as possible

File: mne/time_frequency/tests/test_psd.py [hunk 5]
```diff
@@ -160,6 +160,116 @@ def test_psd():
...
+    assert_allclose(psds_mean, psds_unagg.mean(axis=-1))
+
+    # SciPy's welch() function corrects the median PSD for its bias relative to
+    # the mean.
+    from scipy.signal.spectral import _median_bias
+    median_bias = _median_bias(psds_unagg.shape[-1])
+    assert_allclose(psds_median, np.median(psds_unagg, axis=-1) / median_bias)
+
+
+def test_psd_welch_average_kwarg_epo():
```
Comment (agramfort) [inline mne/time_frequency/tests/test_psd.py]: +1 to merge the 2 tests as @larsoner suggested to limit code duplication

**Comment by hoechenberger** [comment]
@agramfort @larsoner Thanks, will ping you when I'm ready.

File: mne/time_frequency/tests/test_psd.py (hunk 4 above)
Comment (hoechenberger) [reply mne/time_frequency/tests/test_psd.py]: I've changed this to `tmin, tmax = -0.5, 0.5`; the entire test (i.e. sum of runs for all three parametrizations `raw`, `epochs`, `evoked`) now finishes in less than 800 ms on my laptop computer.

File: mne/io/array/tests/test_array.py [hunk 6]
```diff
@@ -71,7 +71,7 @@ def test_array_copy():
 
 
 @pytest.mark.slowtest
-@requires_version('scipy', '0.12')
+@requires_version('scipy', '1.2.0')
```
Comment (agramfort) [inline mne/io/array/tests/test_array.py]: @hoechenberger you should not have to bump requirements for an option that is not specifically tested here. In other words scipy1.2 should only become a requirement if you set average different from 'mean' is the psd_welch function.

File: tutorials/time-freq/plot_sensors_time_frequency.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
 """  # noqa: E501
 # Authors: Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
 #          Stefan Appelhoff <stefan.appelhoff@mailbox.org>
+#          Richard Höchenberger <richard.hoechenberger@gmaill.com>
```
Comment (agramfort) [inline tutorials/time-freq/plot_sensors_time_frequency.py]:
```suggestion
#          Richard Höchenberger <richard.hoechenberger@gmail.com>
```

I suppose

**Comment by larsoner** [comment]
@hoechenberger if you rebase it should fix the Travis error

File: mne/io/array/tests/test_array.py (hunk 6 above)
Comment (hoechenberger) [reply mne/io/array/tests/test_array.py]: I had to bump requirements for all tests that invoke `psd_welch()` as my current implementation would _always_ pass the `average` keyword down to `scipy.signal.welch()`. This was probably not the best idea. I will revert adding all these `requires_version` decorators in ad3063a, and add special handling for `average='mean'` to retain backward-compat, i.e., will do sth that effectively boils down to this:
```python
if average == 'mean':
    scipy.signal.welch(...)
elif average == 'median':
    scipy.signal.welch(, average='median')
```

File: tutorials/time-freq/plot_sensors_time_frequency.py [hunk 8]
```diff
@@ -82,6 +83,52 @@
...
+# bias relative to the mean), which is a more robust measure.
+
+# Estimate PSDs based on "mean" and "median" averaging for comparison.
+kwargs = dict(fmin=2, fmax=40, n_jobs=1)
+psds_welch_mean, freqs_mean = psd_welch(epochs, average='mean', **kwargs)
+psds_welch_median, freqs_median = psd_welch(epochs, average='median', **kwargs)
+
+# The returned frequencies should be identical, since they were estimated from
+# the same input data.
+assert np.array_equal(freqs_mean, freqs_median)
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply tutorials/time-freq/plot_sensors_time_frequency.py]: I included it to demonstrate that the sample frequencies are identical in both cases; I do agree though that an `assert` statement most probably doesn't belong there. Do you think changing it to a simple
```python
print(np.array_equal(freqs_mean, freqs_median))
```
would do the trick? Or would you rather I remove this part entirely?

**Comment by agramfort** [comment]
Remove it

**Comment by agramfort** [comment]
great job @hoechenberger !

**Comment by hoechenberger** [comment]
Thanks @agramfort!

Just got me thinking, what do we do about the visualization functions? e.g., `viz.raw.plot_raw_psd()` currently has the following signature:
https://github.com/mne-tools/mne-python/blob/88a2e3dae4b22ca13d088db8b2213745cd9a653c/mne/viz/raw.py#L562-L567

The `average` kwarg there refers to averaging _across channels_, so now there is some kind of inconsistency in kwarg meaning between `raw_plot_psd()` and `psd_wech()`. I wonder if it would make sense to
- deprecate the `average` keyword in the PSD plotting function,
- replace it with a new a keyword `average_channels`, and
- add a new keyword `average_segments` to allow switching from `mean` to `median` averaging

**Comment by agramfort** [comment]
true...

maybe we can deprecate average=bool to make it a string by default mean or
median ?

I personally never use average=True as i use this function for spotting bad
channels

>

**Comment by larsoner** [comment]
We have a number of uses of `average`:
```
$ git grep "average : "
connectivity/spectral.py:    faverage : boolean
decoding/csp.py:        average : float | None
decoding/csp.py:        average : float | None
decoding/transformer.py:    average : bool, default False
evoked.py:    grand_average : Evoked
preprocessing/ssp.py:    average : bool
preprocessing/ssp.py:    average : bool
time_frequency/psd.py:    average : str | None
time_frequency/psd.py:    average : str | None
time_frequency/tfr.py:    average : bool, default True
time_frequency/tfr.py:    average : bool, default True
utils/docs.py:average : bool
utils/numerics.py:    grand_average : Evoked | AverageTFR
viz/topomap.py:    average : float | None

$ git grep "plot_psd_average"
utils/docs.py:docdict["plot_psd_average"] = """
viz/epochs.py:    %(plot_psd_average)s
viz/raw.py:    %(plot_psd_average)s
```
It almost always means "average across epochs". I'd rather not have to deprecate all of these, though. Maybe the first step is just to make sure it's clear in each instance what the average is taken over, and maybe that will end up being good enough.

**Comment by hoechenberger** [comment]
@larsoner I still believe it would be good to expose the new `average='median'` functionality of `psd_welch()` to users of `*_plot_psd()` too. One could retain all existing kwargs and introduce a new one, like `average_segments`. WDYT?

**Comment by mmagnuski** [comment]
@hoechenberger does this PR change the default averaging behaviour? Previously it was `np.nanmean` which was useful when one of many segments was rejected via `reject_by_annotations` - now I get all nans for some of my data that previously were all good.

**Comment by mmagnuski** [comment]
I checked: yes, it seems it does as scipy's welch uses `np.mean` not `np.nanmean`. To revert to previous behaviour we would have to always use average=None of scipy's welch and do the reduction on our side.

**Comment by larsoner** [comment]
@mmagnuski can you make a PR for a quick fix?

**Comment by mmagnuski** [comment]
sure, but on friday most likely (I'm travelling now).

**Pull request merged**